pkg/lokifrontend/frontend/v2/frontend_test.go

```diff
@@ -252,6 +251,94 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	})
 }
 
+func TestFrontendStoppingWaitsForEmptyInflightRequests(t *testing.T) {
+	delayResponse := 10 * time.Millisecond
+	f, _ := setupFrontend(t, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		// We cannot call QueryResult directly, as Frontend is not yet waiting for the response.
+		// It first needs to be told that enqueuing has succeeded.
+		go sendResponseWithDelay(f, 2*delayResponse, "test", msg.QueryID, &httpgrpc.HTTPResponse{
+			Code: 200,
+			Body: []byte("ok"),
+		})
+
+		// give enough time to fill up inflight requests
+		time.Sleep(delayResponse)
+		return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.OK}
+	})
+
+	inflightRequests := 10
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for i := 0; i < inflightRequests; i++ {
+		go func() {
+			_, err := f.RoundTripGRPC(user.InjectOrgID(ctx, "test"), &httpgrpc.HTTPRequest{})
+			require.NoError(t, err)
+		}()
+	}
+
+	require.Eventually(t, func() bool {
+		return f.requests.count() == inflightRequests
+	},
+		3*delayResponse, // wait at least 2*delayResponse to wait for queries to be finished and removed from inlight requests map
```

Comment:
nit:
it's too short... I would make it at least 1s ... it would prevent us from having flaky tests ;) 
if 10 requests are executed sequentially (it is unbelievable but possible;))) ), then the test will take > 200ms.... and this check will be failed.

Reply:
These 10 requests are executed concurrently.

Reply:
yep, I saw ;) anyway, the check happens every tick and completes as soon the function returns true, so, making timeout bigger just removed the possibility of this test to became flaky test ;) not critical ;)